tests/unit/test_project.py

```diff
@@ -76,139 +112,278 @@ def test_project_unmarshal(yaml_data):
+
+    mock_platforms = {"mock": {"build-on": [other_arch], "build-for": "amd64"}}
+    assert (
+        f"must be built on one of the following architectures: {[other_arch]}"
+        in reload_project_platforms(mock_platforms)
+    )
+
+
 @pytest.mark.parametrize(
     "field", ["name", "version", "base", "parts", "description", "summary", "license"]
 )
-def test_project_missing_field(yaml_data, field):
-    del yaml_data[field]
+def test_project_missing_field(yaml_loaded_data, field):
+    del yaml_loaded_data[field]
 
     with pytest.raises(ProjectValidationError) as err:
-        Project.unmarshal(yaml_data)
+        Project.unmarshal(yaml_loaded_data)
     assert str(err.value) == (
         "Bad rockcraft.yaml content:\n"
         f"- field '{field}' required in top-level configuration"
     )
 
 
-def test_project_extra_field(yaml_data):
-    yaml_data["extra"] = "invalid"
+def test_project_extra_field(yaml_loaded_data):
+    yaml_loaded_data["extra"] = "invalid"
 
     with pytest.raises(ProjectValidationError) as err:
-        Project.unmarshal(yaml_data)
+        Project.unmarshal(yaml_loaded_data)
     assert str(err.value) == (
         "Bad rockcraft.yaml content:\n"
         "- extra field 'extra' not permitted in top-level configuration"
     )
 
 
-def test_project_parts_validation(yaml_data):
-    yaml_data["parts"]["foo"]["invalid"] = True
+def test_project_parts_validation(yaml_loaded_data):
+    yaml_loaded_data["parts"]["foo"]["invalid"] = True
 
     with pytest.raises(ProjectValidationError) as err:
-        Project.unmarshal(yaml_data)
+        Project.unmarshal(yaml_loaded_data)
     assert str(err.value) == (
         "Bad rockcraft.yaml content:\n"
         "- extra field 'invalid' not permitted in 'parts.foo' configuration"
     )
 
 
-def test_project_load(yaml_data, tmp_path):
-    rockcraft_file = tmp_path / "rockcraft.yaml"
+def test_project_load(yaml_data, yaml_loaded_data):
+    rockcraft_file = Path("rockcraft.yaml")
     rockcraft_file.write_text(
-        textwrap.dedent(json.dumps(yaml_data)),
+        yaml_data,
         encoding="utf-8",
     )
 
     project = load_project(str(rockcraft_file))
 
-    for attr, v in yaml_data.items():
+    for attr, v in yaml_loaded_data.items():
         if attr == "license":
             # the var license is a built-in,
             # so we workaround it by using an alias
             attr = "rock_license"
+        if attr == "platforms":
+            # platforms get mutated at validation time
+            assert project.__getattribute__(attr).keys() == v.keys()
+            assert all(
+                "build_on" in platform
+                for platform in project.__getattribute__(attr).values()
+            )
+            assert all(
+                "build_for" in platform
+                for platform in project.__getattribute__(attr).values()
+            )
+            continue
         assert project.__getattribute__(attr) == v
 
+    rockcraft_file.unlink()
```

Comment:
I think we should either wrap the whole test in a `try/finally` to ensure this cleanup, or maybe use `request.addfinalizer()` 
(or with the `tmp_path` fixture I guess we wouldn't have to worry about this test file).

Reply:
`tmp_path` is back in there again